src/ngx_http_lua_log_ringbuf.c

```diff
@@ -0,0 +1,206 @@
+ * placehold data
+ */
+
+static void *
+ngx_http_lua_log_ringbuf_next_header(ngx_http_lua_log_ringbuf_t *rb)
+{
+    ngx_http_lua_log_ringbuf_header_t       *head;
+
+    /* useless data */
+    if (rb->size - (rb->head - rb->data) <
+        sizeof(ngx_http_lua_log_ringbuf_header_t))
+    {
+        rb->head = rb->data;
+        return rb->head;
+    }
+
+    /* placehold data */
+    head = (ngx_http_lua_log_ringbuf_header_t *) rb->head;
+    if (!head->is_data) {
+        rb->head = rb->data;
+        return rb->head;
+    }
+
+    return rb->head;
+}
+
+
+/* append data to ring buffer directly */
+
+static void
+ngx_http_lua_log_ringbuf_append(ngx_http_lua_log_ringbuf_t *rb,
+    int is_data, int log_level, void *buf, int n)
+{
+    ngx_http_lua_log_ringbuf_header_t        *head;
+
+    head = (ngx_http_lua_log_ringbuf_header_t *) rb->tail;
+    head->len = n;
+    head->log_level = log_level;
+    head->is_data = is_data;
+
+    if (!is_data) {
+        rb->tail = rb->data;
+        return;
+    }
+
+    rb->tail += sizeof(ngx_http_lua_log_ringbuf_header_t);
+    ngx_memcpy(rb->tail, buf, n);
+    rb->tail += n;
+    rb->count++;
+
+    return;
+}
+
+
+/* size of free spaces */
+
+static size_t
+ngx_http_lua_log_ringbuf_free_spaces(ngx_http_lua_log_ringbuf_t *rb)
+{
+    if (rb->tail == rb->head && rb->tail == rb->data) {
+        return rb->size;
+    }
+
+    if (rb->tail > rb->head) {
+        return rb->data + rb->size - rb->tail;
+    }
+
+    return rb->head - rb->tail;
+}
+
+
+/*
+ * try to write log data to ring buffer, throw away old data
+ * if there was not enough free spaces.
+ */
+
+ngx_int_t
+ngx_http_lua_log_ringbuf_write(ngx_http_lua_log_ringbuf_t *rb, int log_level,
+    void *buf, size_t n)
+{
+    size_t              free_spaces, head_len;
+
+    ngx_http_lua_log_ringbuf_header_t       *head;
+
+    head_len = sizeof(ngx_http_lua_log_ringbuf_header_t);
+
+    if (n + head_len > rb->size) {
+        return NGX_ERROR;
+    }
+
+    free_spaces = ngx_http_lua_log_ringbuf_free_spaces(rb);
+
+    if (free_spaces < n + head_len) {
+        /* if the right space is not enough, mark it as placehold data */
+        if ((size_t)(rb->data + rb->size - rb->tail) < n + head_len) {
+            ngx_http_lua_log_ringbuf_append(rb, 0, 0, NULL, 0);
+        }
+
+        do {    /*  throw away old data */
+            if (rb->head != ngx_http_lua_log_ringbuf_next_header(rb)) {
+                break;
+            }
+
+            head = (ngx_http_lua_log_ringbuf_header_t *) rb->head;
+            rb->head += head_len + head->len;
+            rb->count--;
+
+            ngx_http_lua_log_ringbuf_next_header(rb);
+            free_spaces = ngx_http_lua_log_ringbuf_free_spaces(rb);
+        } while (free_spaces < n + head_len);
+    }
+
+    ngx_http_lua_log_ringbuf_append(rb, 1, log_level, buf, n);
+
+    return NGX_OK;
+}
+
+
+/* read log from ring buffer, do reset if all of the logs were readed. */
+
```

Comment:
style: no blank line here. Please fix other similar places accordingly. Thank you.